src/realm/node_directory.h

```diff
@@ -0,0 +1,105 @@
+#ifndef NODE_DIRECTORY_H
+#define NODE_DIRECTORY_H
+
+#include <atomic>
+#include <cstdint>
+#include <unordered_map>
+#include <vector>
+#include <shared_mutex>
+
+#include "realm/nodeset.h"
+#include "realm/event.h"
+#include "realm/nodeset.h"
+
+namespace Realm {
+  // -----------------------------------------------------------------
+  // Small immutable header describing a peer
+  // ------------------------------------------------------------------
+  struct NodeMeta {
+    NodeID id{NodeID(-1)};
+    uint64_t epoch{0};
+    uint32_t ip{0};
+    uint16_t udp_port{0};
+    uint8_t flags{0};
+    std::vector<uint8_t> worker_address;
+    uint32_t dev_index{0};
+
+    std::vector<uint8_t> machine_model;
+  };
+
+  struct NodeSlot {
+    NodeMeta brief;
+    std::atomic<uint8_t> state{0}; // 0 live, 1 retiring, 2 failed
+  };
+
+  // TODO:
+  // 1. It needs to be fast, concurent, scalable
+  // 2. It needs to integrate well with custom-built DHT
+  // and other external membership solutions such as etcd, zookeeper
+  // 3. Interface and implementation needs to be cleaned up
+
+  class NodeDirectory {
```

Comment:
Could you please add some comments to the member function of NodeDirectory to explain what they are and when they need to be called.